kotlin/navigate-route/src/main/java/com/esri/arcgisruntime/sample/navigateroute/MainActivity.kt

```diff
@@ -0,0 +1,305 @@
+import com.esri.arcgisruntime.mapping.ArcGISMap
+import com.esri.arcgisruntime.mapping.Basemap
+import com.esri.arcgisruntime.mapping.Viewpoint
+import com.esri.arcgisruntime.mapping.view.Graphic
+import com.esri.arcgisruntime.mapping.view.GraphicsOverlay
+import com.esri.arcgisruntime.mapping.view.LocationDisplay
+import com.esri.arcgisruntime.navigation.DestinationStatus
+import com.esri.arcgisruntime.navigation.RouteTracker
+import com.esri.arcgisruntime.navigation.TrackingStatus
+import com.esri.arcgisruntime.symbology.SimpleLineSymbol
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteParameters
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteResult
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteTask
+import com.esri.arcgisruntime.tasks.networkanalysis.Stop
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.android.synthetic.main.layout_navigation_controls.*
+import java.util.concurrent.ExecutionException
+
+class MainActivity : AppCompatActivity() {
+
+  private var textToSpeech: TextToSpeech? = null
+
+  private var isTextToSpeechInitialized = false
+
+  companion object {
+    private val TAG: String = this::class.java.simpleName
+
+    /**
+     * Creates a list of stops along a route.
+     */
+    private fun getStops(): List<Stop> {
+      // San Diego Convention Center
+      val conventionCenter = Stop(Point(-117.160386, 32.706608, SpatialReferences.getWgs84()))
+      // USS San Diego Memorial
+      val memorial = Stop(Point(-117.173034, 32.712327, SpatialReferences.getWgs84()))
+      // RH Fleet Aerospace Museum
+      val aerospaceMuseum = Stop(Point(-117.147230, 32.730467, SpatialReferences.getWgs84()))
+
+      return listOf(conventionCenter, memorial, aerospaceMuseum)
+    }
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // create a map and set it to the map view
+    mapView.map = ArcGISMap(Basemap.createStreetsVector())
+
+    // create a graphics overlay to hold our route graphics and clear any graphics
+    val graphicsOverlay = GraphicsOverlay()
+    mapView.graphicsOverlays.add(graphicsOverlay).also { graphicsOverlay.graphics.clear() }
+
+    // create text-to-speech to replay navigation voice guidance
+    textToSpeech = TextToSpeech(this) { status ->
+      if (status != TextToSpeech.ERROR) {
+        textToSpeech?.language = Resources.getSystem()
+          .configuration.locale
+        isTextToSpeechInitialized = true
+      }
+    }
+
+    // generate a route with directions and stops for navigation
+    val routeTask = RouteTask(this, getString(R.string.routing_service_url))
+    val routeParametersFuture = routeTask.createDefaultParametersAsync()
+    routeParametersFuture.addDoneListener {
+      try {
+        // define the route parameters
+        val routeParameters = routeParametersFuture.get().also {
```

Comment:
probably apply is better here since you're effectively setting more properties of the route parameters


also is good for doing something with the receiver like updating the UI

Reply:
I believe I was going back and forth between also and let because in order to use getStops() I needed this to be MainActivity. However with getStops() in flux I've changed it to apply now.